### PR TITLE
WIP - migration to lsp-protocol.el and preparation to plists

### PR DESCRIPTION
**Pull request by yyoncho**
https://github.com/emacs-lsp/lsp-mode/issues/1751

### CODE REVIEW

File: test/lsp-protocol-test.el [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+;;; Code:
+
+
+(require 'lsp-protocol)
+(require 'ert)
+
+(lsp-interface (MyPosition (:line :character :camelCase) (:optional)))
+
+(ert-deftest lsp-test-lsp-interface ()
```
Comment (ericdallo) [inline test/lsp-protocol-test.el:33]: :clap:

File: lsp-protocol.el [hunk 2]
```diff
@@ -55,46 +64,101 @@ Example usage with `dash`.
...
+                      ((ht? object)
                        (-all? (lambda (prop)
-                                (plist-get object prop))
-                              ',required)))
+                                (ht-get object prop))
+                              ',(-map (lambda (field-name)
+                                        (substring (symbol-name field-name) 1))
+                                      required)))
+                      ((listp object) (-all? (lambda (prop)
+                                               (plist-get object prop))
```
Comment (kiennq) [inline lsp-protocol.el:104]: This should be `plist-member`, else it will return `nil` for `nil` member

Comment (kiennq) [reply lsp-protocol.el:104]:
```suggestion
                                               (plist-member object prop))
```

File: lsp-protocol.el [hunk 3]
```diff
@@ -55,46 +63,100 @@ Example usage with `dash`.
...
-                     (unless (member key ',(-map #'cl-first params))
+                     (unless (or (member key ',(-map #'cl-first params))
+                                 (s-starts-with? ":_" (symbol-name key)))
                        (error "Unknown key: %s. Available keys: %s" key ',(-map #'cl-first params)))
-                     `(gethash ,(substring (symbol-name
-                                            (cl-rest (assoc key ',params)))
-                                           1)
-                               ,source))
+                     ,(if lsp-use-plists
+                          ``(plist-get ,source
```
Comment (kiennq) [inline lsp-protocol.el:70]: What is double back quote for? I'm asking out of curious

File: lsp-mode.el [hunk 4]
```diff
@@ -26,6 +26,7 @@
 ;; Emacs client/library for the Language Server Protocol
 
 ;;; Code:
+(setq lsp-plists nil)
```
Comment (kiennq) [inline lsp-mode.el:29]: This is not needed.

File: lsp-protocol.el (hunk 3 above)
Comment (kiennq) [inline lsp-protocol.el:70]: This one should also be `plist-member`

File: lsp-protocol.el [hunk 5]
```diff
@@ -55,46 +63,100 @@ Example usage with `dash`.
...
+                                       ,(if (s-starts-with? ":_" (symbol-name key))
+                                            key
+                                          (cl-rest (assoc key ',params))))
+                        ``(gethash ,(if (s-starts-with? ":_" (symbol-name key))
```
Comment (kiennq) [inline lsp-protocol.el:74]: Do you have something similar to `plist-member` for `gethash` ?

File: lsp-protocol.el [hunk 6]
```diff
@@ -55,46 +63,100 @@ Example usage with `dash`.
...
                   `(defun ,(intern (format "dash-expand:&%s?" interface)) (key source)
                      (unless (member key ',(-map #'cl-first params))
                        (error "Unknown key: %s. Available keys: %s" key ',(-map #'cl-first params)))
-                     `(when ,source
-                        (gethash ,(substring (symbol-name
-                                              (cl-rest (assoc key ',params)))
-                                             1)
-                                 ,source)))
+                     ,(if lsp-use-plists
+                          ``(plist-member ,source
```
Comment (kiennq) [inline lsp-protocol.el:84]: This should be `plist-get`

**Comment by kiennq** [comment]
@yyoncho @ericdallo Should we add another commit to set `lsp-use-plists` to nil and process to merge this branch?
It contains quite a lot `lsp-protocol` improvements and there're external repo that depends on it.

**Comment by yyoncho** [comment]
> @yyoncho @ericdallo Should we add another commit to set `lsp-use-plists` to nil and process to merge this branch?
> It contains quite a lot `lsp-protocol` improvements and there're external repo that depends on it.

Yes, I guess after some testing.

**Pull request merged**